File: Shamus/MazeBuilder.cpp
// MazeBuilder.cpp: implementation of the CMazeBuilder class.
//
//////////////////////////////////////////////////////////////////////

#include "MazeBuilder.h"

//////////////////////////////////////////////////////////////////////
// Construction/Destruction
//////////////////////////////////////////////////////////////////////

CMazeBuilder::CMazeBuilder()
{
	_currentMaze = nullptr;
}

CMazeBuilder::~CMazeBuilder()
{
	delete _currentMaze;
}

void CMazeBuilder::BuildMaze(const unsigned char *data)
{
	_currentMaze = new CMaze;
	int idx = 0;
	for (signed char index1 = 0;index1 < MAX_ROOM_X;index1++)
	{
		for (signed char index2 = 0;index2 < MAX_ROOM_Y;index2++)
		{
			const auto room = new CRoom(XCOUNT, YCOUNT);
			for (signed char i = 0;i < XCOUNT;i++)
			{
				for (signed char j = 0;j < YCOUNT;j++)
				{
					const unsigned char item = data[idx];
					if (item > 0)
					{
						room->SetItem(i, YCOUNT - j - 1, item);
					}
					idx++;
				}
			}
			_currentMaze->AddRoom(index1, index2, room);
		}
	}
}

CMaze* CMazeBuilder::GetMaze() const
{
	return _currentMaze;
}

File: Shamus.LevelEditor/MazeBuilder.cs
using System;

namespace Shamus.LevelEditor
{
    public class MazeBuilder
    {
        private Maze _currentMaze;

        public byte[] Data(Maze maze)
        {
            const int length = Config.DATA_SIZE;
            byte[] data = new byte[length];
            int position = 0;
            for (int index1 = 0; index1 < Config.MAX_ROOM_X; index1++)
            {
                for (int index2 = 0; index2 < Config.MAX_ROOM_Y; index2++)
                {
                    for (int i = 0; i < Config.XCOUNT; i++)
                    {
                        for (int j = 0; j < Config.YCOUNT; j++)
                        {
                            data[position] = (byte)maze.GetObject(index1, index2, i, j);
                            position++;
                        }
                    }
                }
            }

            return data;
        }

        public Maze CurrentMaze => _currentMaze;

        public void BuildMaze(byte[] data)
        {
            int position = 0;
            _currentMaze = new Maze();
            for (int index1 = 0; index1 < Config.MAX_ROOM_X; index1++)
            {
                for (int index2 = 0; index2 < Config.MAX_ROOM_Y; index2++)
                {
                    for (int i = 0; i < Config.XCOUNT; i++)
                    {
                        for (int j = 0; j < Config.YCOUNT; j++)
                        {
                            _currentMaze.SetObject(index1, index2, i, j, (Item)data[position]);
                            position++;
                        }
                    }
                }
            }
        }
    }
}
